### PR TITLE
Adding the Fishbyt app

### PR DESCRIPTION
**Pull request by vlauffer**
The fish facts app that everybody needed but nobody wanted (unless you are David Winitch, in which case you wanted this app as well) is finally here. This is my first time doing a pull request so hopefully everything checks out

fishbyt
.

### CODE REVIEW

**Comment by vlauffer** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by vlauffer** [comment]
recheck

File: apps/fishbyt/fishbyt.star [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+    #get fish data and pick a random fish
+    fish_barrel = get_fish_barrel()
+    random_index=random(len(fish_barrel))
+    fish = fish_barrel[random_index]
+
+    #check and set an available fish picture. If no picture is found, display FAIL_IMAGE
+    fish_pic = fish["Species Illustration Photo"]["src"]
+    if fish_pic==None: 
+        print("No fish in this pond")
```
Comment (betterengineering) [inline apps/fishbyt/fishbyt.star:40]:
```suggestion
```

File: apps/fishbyt/fishbyt.star [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        fish_pic=base64.decode(FAIL_IMAGE)
+    else:    
+        print("Look at that fish!")
```
Comment (betterengineering) [inline apps/fishbyt/fishbyt.star:43]:
```suggestion
```

File: apps/fishbyt/fishbyt.star [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        ),
+    )
+
+
+def get_fish_pic(url):
+    key = base64.encode(url)
+    fish_pic_cache = cache.get(key)
+
+    if fish_pic_cache != None:
+        print("Caught one!")
```
Comment (betterengineering) [inline apps/fishbyt/fishbyt.star:108]:
```suggestion
```

File: apps/fishbyt/fishbyt.star [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        return base64.decode(fish_pic_cache)
+
+    res = http.get(url = url)
+    if res.status_code != 200:
+        fail("No fish here! Request to %s failed with status code: %d - %s" % (url, res.status_code, res.body()))
+
+    print("Let's catch this fish!")
```
Comment (betterengineering) [inline apps/fishbyt/fishbyt.star:115]:
```suggestion
```

File: apps/fishbyt/fishbyt.star [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    cache.set(key, base64.encode(res.body()), ttl_seconds = CACHE_TTL_SECONDS)
+    return res.body()
+
+
+def get_fish_barrel():
+
+    fish_barrel_cached = cache.get("fish_barrel")
+    fish_barrel=[]
+    if fish_barrel_cached != None:
+        print("My barrel is full of fish!")
```
Comment (betterengineering) [inline apps/fishbyt/fishbyt.star:125]:
```suggestion
```

File: apps/fishbyt/fishbyt.star [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        fish_barrel = json.decode(fish_barrel_cached)
+
+    else:
+        print("Let's go fishing!")
```
Comment (betterengineering) [inline apps/fishbyt/fishbyt.star:129]:
```suggestion
```

File: apps/fishbyt/fishbyt.star [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+            return [random_attribute, fish_fact_sin_chars]
+        
+        #pick random fact from line
+        fish_fact = pick_random_lines(fish_facts_split)
+        if fish_fact == "no_valid_lines":
+            continue
+
+        return [random_attribute, fish_fact]
+
+    print("No facts found in get_fact")
```
Comment (betterengineering) [inline apps/fishbyt/fishbyt.star:171]:
```suggestion
```

File: apps/fishbyt/fishbyt.star [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    return ["error", "no facts found"]
+
+def pick_random_lines(facts):
+    #find a random fact from the array of split lines and return it
+    for i in range(0, len(facts)-1):
+        random_fact_index = random(len(facts))
+        random_fact = facts.pop(random_fact_index)
+        random_fact = remove_chars(random_fact)
+        if random_fact=="":
+            print("blank fact found, picking another line")
```
Comment (betterengineering) [inline apps/fishbyt/fishbyt.star:181]:
```suggestion
```

File: apps/fishbyt/fishbyt.star [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+            continue
+
+        if not fact_length_check_by_char(random_fact):
+            print("Fact is too long: "+ random_fact)
```
Comment (betterengineering) [inline apps/fishbyt/fishbyt.star:185]:
```suggestion
```

File: apps/fishbyt/fishbyt.star [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        #If just one fact is available, return if not empty or too long
+        fish_facts_split = fish_facts.split("</li>")
+        if len(fish_facts_split) ==1:
+            fish_fact_sin_chars = remove_chars(fish_facts_split[0])
+            if fish_facts_split[0] =="":
+                print("2: No facts at " + random_attribute)
+                continue
+
+            if not fact_length_check_by_char(fish_fact_sin_chars):
+                print("Fact is too long: "+fish_fact_sin_chars)
```
Comment (betterengineering) [inline apps/fishbyt/fishbyt.star:159]:
```suggestion
```

File: apps/fishbyt/fishbyt.star [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+            print("1: No facts at " + random_attribute)
+            continue
+
+        #checks to see if there are multiple facts available. If so, pick a random line
+        #If just one fact is available, return if not empty or too long
+        fish_facts_split = fish_facts.split("</li>")
+        if len(fish_facts_split) ==1:
+            fish_fact_sin_chars = remove_chars(fish_facts_split[0])
+            if fish_facts_split[0] =="":
+                print("2: No facts at " + random_attribute)
```
Comment (betterengineering) [inline apps/fishbyt/fishbyt.star:155]:
```suggestion
```

File: apps/fishbyt/fishbyt.star [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+def get_fact(fish_info, content_titles_clone):
+    #find a random fact
+    for i in (range(0, len(content_titles_clone))):
+        random_index = random(len(content_titles_clone))
+        random_attribute = content_titles_clone.pop(random_index)
+        fish_facts = fish_info[random_attribute]
+
+        if not fish_facts:
+            print("1: No facts at " + random_attribute)
```
Comment (betterengineering) [inline apps/fishbyt/fishbyt.star:146]:
```suggestion
```

**Review by betterengineering — APPROVED** [review]
Thanks for this app @vlauffer! I love it 🐟 . If you run `make lint` it will resolve the failed linter. Your prints are clever and it's painful to ask to remove them (I laughed out loud when I saw them in my terminal) - but the all end up in our server logs 😂

**Comment by vlauffer** [comment]
@betterengineering thanks for the help! And I completely understand about the fish prints, that server log would look too crazy lol. I'm having a small issue w/ buildifier not being on the right path but I'll figure it out and update the PR!

**Pull request merged**